src/arithmetic/algebraic_expression.c

```diff
@@ -305,95 +201,357 @@ void AlgebraicExpression_PrependTerm(AlgebraicExpression *ae, GrB_Matrix m, bool
+    AlgebraicExpressionOperand op;
+    op.free = false;
+    op.diagonal = true;
+    op.transpose = false;
+    op.operand = Node_GetMatrix(n);
+    return op;
+}
+
+static AlgebraicExpressionOperand _AlgebraicExpression_OperandFromEdge (
+    Edge *e,
+    bool transpose,
+    const AST* ast
+) {
+    AlgebraicExpressionOperand op;
+
+    // TODO: find a better way for retriving AST graph entity.
+    GrB_Matrix mat = Edge_GetMatrix(e);
+    AST_LinkEntity *astEdge = (AST_LinkEntity*)MatchClause_GetEntity(ast->matchNode, e->alias);
+    assert(astEdge);
+
+    int labelCount = AST_LinkEntity_LabelCount(astEdge);
+    bool freeMatrix = (labelCount > 1);
+    // [:A|:B]
+    // Create matrix M = A+B.
+    if(labelCount > 1) {
+        GraphContext *gc = GraphContext_GetFromTLS();
+        Graph *g = gc->g;
+
+        GrB_Matrix m;
+        GrB_Matrix_new(&m, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+
+        for(int i = 0; i < labelCount; i++) {
+            char *label = astEdge->labels[i];
+            Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_EDGE);
+            if(!s) continue;
+            GrB_Matrix l = Graph_GetRelationMatrix(g, s->id);
+            GrB_Info info = GrB_eWiseAdd_Matrix_Semiring(m, NULL, NULL, Rg_structured_bool, m, l, NULL);
         }
+        mat = m;
+    }
 
-        // ()-[:A|:B.]->()
-        // Create matrix M, where M = A+B+...
-        int labelCount = AST_LinkEntity_LabelCount(astEdge);
-        if(labelCount > 1) {
-            GraphContext *gc = GraphContext_GetFromTLS();
-            Graph *g = gc->g;
-
-            GrB_Matrix m;
-            GrB_Matrix_new(&m, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
-
-            for(int i = 0; i < labelCount; i++) {
-                char *label = astEdge->labels[i];
-                Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_EDGE);
-                if(!s) continue;
-                GrB_Matrix l = Graph_GetRelationMatrix(g, s->id);
-                GrB_Info info = GrB_eWiseAdd_Matrix_Semiring(m, NULL, NULL, Rg_structured_bool, m, l, NULL);                
+    op.operand = mat;
+    op.diagonal = false;
+    op.free = freeMatrix;
+    op.transpose = transpose;
+    return op;
+}
+
+AlgebraicExpression *AlgebraicExpression_Empty(void) {
+    return _AE_MUL(1);
+}
+
+/* In case edge a and b share a node:
+ * (a)-[E0]->(b)<-[e1]-(c)
+ * than the shared entity is returned
+ * if edges are disjoint, NULL is returned. */
+static Node* _SharedNode(const Edge *a, const Edge *b) {
+    assert(a && b);
+    if(a->src == b->src) return a->src;
+    if(a->src == b->dest) return a->src;
+    if(a->dest == b->src) return a->dest;
+    if(a->dest == b->dest) return a->dest;
+    return NULL;
+}
+
+static AlgebraicExpression* _AlgebraicExpression_From_Path(Edge **path, uint path_len, const AST *ast) {
+    assert(path && path_len > 0 && ast);
```

Comment:
Assertion failure on:
`MATCH (a)-[]->(a)-[]->() RETURN a`
With this query, the second AlgebraicExpression being constructed has a `level` of 0 after the call to `_DeepestLevel` on 518.

I'm not sure how rigorous we need to be right now about edge cases like this. This PR is a big improvement, so I don't really mind merging it and opening an issue to pursue this later.

Reply:
Good catch!